### PR TITLE
Improved documentation for DataFrame.join

### LINKED ISSUE #12188: Confusing interpretation of what DataFrame.join does
From the docs I see that the difference between .merge and .join is that .join operates using indexes by default, but it also lets you use columns,  so I tried to use it for that since it sounds natural when coming from the SQL world.

From the [docs](http://pandas.pydata.org/pandas-docs/version/0.17.1/generated/pandas.DataFrame.join.html):

> on : column name, tuple/list of column names, or array-like
Column(s) to use for joining, otherwise join on index. If multiples columns given, the passed DataFrame must have a MultiIndex. Can pass an array as the join key if not already contained in the calling DataFrame. Like an Excel VLOOKUP operation

From my understanding, if on is absent a join operation is performed on the index, if on is present, it would be reasonable to think that the same operation would be performed.

Having said that:

```python
a = pd.DataFrame({'address' : ["1820 SOME STREET", "32 ANOTHER STREET",
                              "1140 BIG STREET", "20 SMALL STREET",
                              "1090 AVENUE NAME"],
                  'id' : [1,2,3,4,5]
                  })
b = pd.DataFrame({'address' : ["1090 AVENUE NAME",
                              "1140 BIG STREET", "20 SMALL STREET",
                              "1820 SOME STREET", "32 ANOTHER STREET"],
                  'id' : [5,3,4,1,2]})

a.join(b, on='id', lsuffix='_x',  rsuffix='_y', how='left')
```
Output:

```
           address_x  id_x          address_y  id_y
0   1820 SOME STREET     1    1140 BIG STREET     3
1  32 ANOTHER STREET     2    20 SMALL STREET     4
2    1140 BIG STREET     3   1820 SOME STREET     1
3    20 SMALL STREET     4  32 ANOTHER STREET     2
4   1090 AVENUE NAME     5                NaN   NaN
```
And if I add id as index:

```python
a.set_index('id', inplace=True)
b.set_index('id', inplace=True)
a.join(b, lsuffix='_x',  rsuffix='_y', how='left')
```

Output:
```
            address_x          address_y
id
1    1820 SOME STREET   1820 SOME STREET
2   32 ANOTHER STREET  32 ANOTHER STREET
3     1140 BIG STREET    1140 BIG STREET
4     20 SMALL STREET    20 SMALL STREET
5    1090 AVENUE NAME   1090 AVENUE NAME
```

Is that the correct behavior? If yes, I think the documentation is misleading. It took me a lot to find the bug in my code and I ended up using merge since .join works in an unexpected way.

I don't think I'm the [only one](http://stackoverflow.com/questions/10114399/pandas-simple-join-not-working) with this issue, so maybe a change in the documentation would help to clarify how .join works.

### PR DESCRIPTION
**Pull request by edublancas**
closes #12188 

I modified the description in DataFrame.join to make clear the difference with DataFrame.merge, also added examples.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -4318,18 +4318,20 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
-            * left: use calling frame's index
-            * right: use input frame's index
-            * outer: form union of indexes
-            * inner: use intersection of indexes
+            How to handle the operation of the two objects. Default: 'left'
+
+            * left: use calling frame's index or column(s)
+            * right: use other frame's index
+            * outer: form union of calling frame's index or column(s) with
+              other frame's index
```
Comment (jreback) [inline pandas/core/frame.py]: I think this is *more* confusing, as this depends if ``on`` is specified or not (so you can simply say that)

File: pandas/core/frame.py [hunk 2]
```diff
@@ -4343,6 +4345,55 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
+        5  A5    K5  NaN   NaN
+
+        Perform an column-on-index left join
+
+        >>> caller.join(other, on='key', how='left', lsuffix='_l', rsuffix='_r')
+            A key_l    B key_r
+        0  A0    K0  NaN   NaN
+        1  A1    K1  NaN   NaN
+        2  A2    K2  NaN   NaN
+        3  A3    K3  NaN   NaN
```
Comment (jreback) [inline pandas/core/frame.py]: maybe just show this (with a comment)

```
In [12]: caller.join(other.set_index('key'), on='key', how='left', lsuffix='_l', rsuffix='_r')
Out[12]: 
    A key    B
0  A0  K0   B0
1  A1  K1   B1
2  A2  K2   B2
3  A3  K3  NaN
4  A4  K4  NaN
5  A5  K5  NaN
```

File: pandas/core/frame.py (hunk 1 above)
Comment (edublancas) [inline pandas/core/frame.py]: What about this?
```
* left: use calling frame's index (or column if on is specified)
* right: use other frame's index
* outer: form union of calling frame's index (or column if on is specified) with other frame's index
```

**Comment by jreback** [comment]
can you update

**Comment by edublancas** [comment]
Done, sorry for the delay.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -4343,6 +4345,46 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
+
+        >>> other
+            B key
+        0  B0  K0
+        1  B1  K1
+        2  B2  K2
+
+        Perform a left join using caller's key column and other frame's index 
+
+        >>> caller.join(other.set_index('key'), on='key', how='left',
```
Comment (jreback) [inline pandas/core/frame.py]: can you add this same example w/o using ``.set_index`` as well.

**Comment by jreback** [comment]
pls rebase/update

**Comment by edublancas** [comment]
Sorry for the delay, I've been working to meet some deadlines for a project. I'll update in the next few days.

**Comment by jreback** [comment]
can you rebase/update

File: pandas/core/frame.py (hunk 3 above)
Comment (edublancas) [inline pandas/core/frame.py]: Just this example? `caller.join(other, how='left', lsuffix='_l', rsuffix='_r')`

File: doc/source/merging.rst [hunk 4]
```diff
@@ -558,10 +558,8 @@ DataFrame instance method, with the calling DataFrame being implicitly
 considered the left object in the join.
 
 The related ``DataFrame.join`` method, uses ``merge`` internally for the
-index-on-index and index-on-column(s) joins, but *joins on indexes* by default
-rather than trying to join on common columns (the default behavior for
-``merge``). If you are joining on index, you may wish to use ``DataFrame.join``
-to save yourself some typing.
```
Comment (jorisvandenbossche) [inline doc/source/merging.rst]: Why did you shorten this? 
I think the " *joins on indexes* by default" is very useful explanation

File: pandas/core/frame.py (hunk 3 above)
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: @jreback I don't think it is possible to not use `set_index`, as `join` always uses the index of `other` (which is actually really confusing ...)

@edublancas the `lsuffix='_l', rsuffix='_r'` is redundant in this case, so I would leave it out

Comment (edublancas) [inline pandas/core/frame.py]: What about having just these two examples:

Perform a left join using caller's key column and other frame's index 

`caller.join(other.set_index('key'), on='key', how='left')`

Set `key` as the index column on `caller` and `other` and perform an index-on-index join.

`caller.set_index('key').join(other.set_index('key'), how='left')`

File: doc/source/merging.rst (hunk 4 above)
Comment (edublancas) [inline doc/source/merging.rst]: I think the shorter explanation is better:

> index-on-index (by default) and column(s)-on-index join. If you are joining on index only, you may wish to use ``DataFrame.join`` to save yourself some typing.

**Comment by jreback** [comment]
can you update according to comments

**Comment by jreback** [comment]
can you rebase / update?

**Comment by edublancas** [comment]
I think the documentation is clear now. One using the dataframes original indexes and two joining using the key columns, the first one setting key as the index in both and the second one using on.

**Comment by jreback** [comment]
thanks @edublancas nice improvement!

**Pull request closed without merging**